hpke/hpke.go

```diff
@@ -0,0 +1,253 @@
+// Package hpke implements the Hybrid Public Key Encryption (HPKE) as specified
+// by draft-irtf-cfrg-hpke-06.
+//
+// HPKE works for any combination of an asymmetric-key encapsulation mechanism
+// (KEM), a key derivation function (KDF), and an authenticated symmetric-key
+// encryption scheme with additional data (AEAD).
+//
+// Specification in
+// https://www.ietf.org/archive/id/draft-irtf-cfrg-hpke-06.html
+package hpke
+
+import (
+	"errors"
+
+	"github.com/cloudflare/circl/kem"
+)
+
+const versionLabel = "HPKE-06"
+
+// Exporter allows exporting secrets from an HPKE context using a
+// variable-length PRF. Export takes as input a context string expCtx and a
+// desired length (in bytes), and produces a secret derived from the internal
+// exporter secret using the corresponding KDF Expand function.
+type Exporter interface {
+	Export(expCtx []byte, len uint16) []byte
+}
+
+// Sealer encrypts a plaintext using an AEAD encryption. The caller supplies
+// the plaintext and associated data; the nonce is stored internally by the
+// sealer and incremented after each call.
+type Sealer interface {
+	Exporter
+	Seal(pt, aad []byte) (ct []byte, err error)
+}
+
+// Opener decrypts a ciphertext using an AEAD encryption. The caller supplies
+// the ciphertext and associated data; the nonce is stored internally by the
+// Opener and incremented after each call.
+type Opener interface {
+	Exporter
+	Open(ct, aad []byte) (pt []byte, err error)
+}
+
+// modeID represents an HPKE variant.
+type modeID = uint8
+
+const (
+	// modeBase provides hybrid public-key encryption.
+	modeBase modeID = 0x00
+	// modePSK provides hybrid public-key encryption with authentication using a
+	// pre-shared key.
+	modePSK modeID = 0x01
+	// modeAuth provides hybrid public-key encryption with authentication using
+	// the sender's secret key.
+	modeAuth modeID = 0x02
+	// modeAuthPSK provides hybrid public-key encryption with authentication
+	// using both a pre-shared key and an asymmetric key.
+	modeAuthPSK modeID = 0x03
+)
+
+// Suite is an HPKE cipher suite consisting of a KEM, KDF, and AEAD algorithm.
+type Suite struct {
+	KemID  KemID
+	KdfID  KdfID
+	AeadID AeadID
+}
+
+type state struct {
+	Suite
+	modeID modeID
+	skS    kem.PrivateKey
+	pkS    kem.PublicKey
+	psk    []byte
+	pskID  []byte
+	info   []byte
+}
+
+// Sender performs hybrid public-key encryption.
+type Sender struct {
+	state
+	pkR kem.PublicKey
+}
+
+// NewSender creates a Sender with knowledge of the receiver's public-key.
+func (suite Suite) NewSender(pkR kem.PublicKey, info []byte) (*Sender, error) {
+	if !suite.isValid() {
+		return nil, errors.New("invalid suite")
+	}
+	return &Sender{
+		state: state{Suite: suite, info: info},
+		pkR:   pkR,
```

Comment:
As long as we're enforcing the validity of the cipher suite, we should also make sure that `pkR` is the correct type. (Here and below.)

Reply:
great observation, added verification